### ISSUE TITLE
Alexa Skill?

### ISSUE THREAD

**Issue by primexa** [mdhiggins/ESP8266-HTTP-IR-Blaster#71]
Hello,
I have done all the steps required for this project and was hoping for an update on the Alexa Skill. Amazon Alexa app updated a couple days ago and I was hoping to see the IR Controller skill on the Alexa Skills store but does not seem to be active yet as of November 1st :(  Do you have any update as I am sure you are anxious as well?

**Comment by mdhiggins** [comment]
It’s finished and approved by amazon. Just waiting for the official push for all the V3 smart home skills. They told me end of October but haven’t given a specific date and obviously that has passed. It’s all in their hands now.

**Comment by primexa** [comment]
Thank you for the quick response, waiting patiently....

**Comment by mdhiggins** [comment]
Did I sent you a beta invite already?

**Comment by primexa** [comment]
No sir, I have gotten a beta invite. That would be fantastic!

On Nov 1, 2017 3:52 PM, "Michael Higgins" <notifications@github.com> wrote:

Did I sent you a beta invite already?

Sent from my iPhone

> On Nov 1, 2017, at 14:39, primexa <notifications@github.com> wrote:
>
> Thank you for the quick response, waiting patiently....
>
> —
> You are receiving this because you commented.

> Reply to this email directly, view it on GitHub, or mute the thread.
>

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/mdhiggins/ESP8266-HTTP-IR-Blaster/issues/71#issuecomment-341220573>,
or mute the thread

**Comment by primexa** [comment]
Sorry, let me reiterate, no I have NOT gotten an invite :)

**Comment by mdhiggins** [comment]
Send me your Amazon account email address

**Comment by primexa** [comment]
sent...

**Comment by mdhiggins** [comment]
Invite sent

**Comment by Pawand15** [comment]
Hi could you please send me an invite for the skill
Amazon account details: pawan.d@outlook.com

Really a great project.. Already made the hardware!
Just received my echo In India.

**Comment by kickaxe60** [comment]
This works for me.  How do we add channel changing capabilities?

Thanks

**Comment by mdhiggins** [comment]
It’s in development for the next update

**Comment by kickaxe60** [comment]
I would be happy to be a beta tester if you need one.  Note that I couldn't get the Input Controller working, no matter what label I gave the input.  I could switch inputs by tricking Alexa by assigning the IR code for input switching using a Playback Controller instead though.

**Comment by mdhiggins** [comment]
That’s probably more of an issue on Alexa’s end. I too had trouble getting that to work at first. It’s only very specific input names and it can be picky about the device name. I use it to change my speaker system around and the following commands work for me:

Alexa change surround sound input to Bluetooth 
Alexa change surround sound input to TV

If it’s still not working I could recommend using the scene controller to make a scene that changes the input. That scene can then be used with a routine to let you say whatever you want.

**Comment by kickaxe60** [comment]
OK I'll try some variations of what you suggest for input switching.  Do you have any info on how to create a scene.  Can you have more that one or can one scene have many triggers?

Thanks

**Comment by mdhiggins** [comment]
You can create many scenes. The scene controller setup options are at the bottom of  the page (tehpsyc.pythonanywhere.com)

**Comment by kickaxe60** [comment]
I can't seem to figure out how to use scenes.  I can enable and put command activate/deactivate commands but that's it.  I click reference but can't see any information on how to use scenes...

**Comment by mdhiggins** [comment]
Just add the URL like you would any other command, update and save, then rerun device discovery

Now you can say “turn on SCENENAME” and “turn off SCENENAME”, or saying “activate SCENENAME” or “deactivate SCENENAME”

Scenes can then also be incorporated into routines

**Comment by kickaxe60** [comment]
So can you can have just one scene per device, and all other controller info e.g. power/on off, step volume... are ignored for that device.  Is that correct? If not, how do I add more than one scene for a particular device? Basically, I would like to add commands like "Alexa, switch TV to CNN"...  I'm just trying to wrap my head around how scenes work.  

Thanks for your help Michael, it's appreciated.

Denis

**Comment by mdhiggins** [comment]
Treat the scene as its own device. You can still set up all the other functionality just do it separately. (The URLs can point to the same ESP8266, that’s fine). 
So you would have your TV with its power and volume control as one device, then just create a scene and name it HDMI1, then another scene and name it Bluray, etc

For things like changing channel to CNN, that is something that I’m working on support for via the channelcontroller but is still in development, so it should be natively supported soon

**Comment by kickaxe60** [comment]
Got it!  Is there a limit as to how many devices Alexa will discover i.e. can I have 20 scenes?

**Comment by mdhiggins** [comment]
I’m not sure, I don’t think there is, there’s a character limit for device discover packet size but its crazy high and you probably won’t hit it without hundreds of items

**Comment by kickaxe60** [comment]
OK Thanks Michael.

**Comment by kickaxe60** [comment]
Hi Michael, I added many scenes yesterday and all work fine.  Today, I want
to add a few more and discovery doesn't find anything and all the other
scenes say offline, although they still work.  Any idea what would cause
that and how I woud go about getting Alexa to discover my new scenes
created through IR Controller?

Thanks

Denis

**Comment by mdhiggins** [comment]
I pushed some security updates and appear to have created a bug not revealed with my initial testing. At work but will fix or roll back those changes tonight

**Comment by mdhiggins** [comment]
Should be resolved now. Just an issue with one of the security tokens

**Comment by kickaxe60** [comment]
Hi Michael, still no luck here. I tried disabling and re-enabling the IR
Controller skill, sign out and back in again and they still show offline.
I have a smart plug which is discovered no problem but none of the IR
Controller virtual devices or scenes get picked up.  I'm away for a few
weeks on holiday and will try again when I get back.

Thanks for your help.

Denis

**Comment by kickaxe60** [comment]
Hi Michael, back from holiday and trying this out again.  I see you've
updated the .ino file and added channel capabilities.  I seem to have
everything setup correctly however, when I try to send a code I get the
following in terminal:

Connection received - MSG
Failed security check, signature is too old
Server: 0
Local: 1512667631
MID:

I'm using the amzn1 account at the bottom of the controller setup.  Any
ideas what my issue might be?

Thanks

Denis

**Comment by mdhiggins** [comment]
Looks like it’s unable to connect to the time server to verify your time stamp. This is needed for authentication. 

Make sure you have all that stuff configured and have your DNS settings configured appropriately. That caused me a major headache before with the same issue.

**Comment by mdhiggins** [comment]
Wait looking at that closer it looks like that’s not the case. The local time stamp worked but the server did not. Are you sending the command from a regular URL or from Alexa? Looks like the server time stamp was not included? Once the user_id is set the device goes into Alexa only mode and won’t accept regular URLs anymore.

**Comment by kickaxe60** [comment]
Thanks, got the basic stuff working again.  I am not able to get channels
to work though.  My echo dot says "this device doesn't support that."  I
have to use raw data as my cable box remote is not recognized.  I can get
them working with scenes by firing off the three numerical codes but that
means creating many channel scenes...
Here's my base url with ip x'd out:
http://xxx.xxx.xxx.x:667/json?pass=pass
Here's a sample of the button 1 syntax:
'data':[8996,4454,514,4454,516,2220,514,2190,544,2218,516,2188,538,2226,514,2190,546,2186,542,2222,510,2218,514,2222,514,2190,546,4454,512,4428,546,4426,546,4424,544],'type':'raw’,’length’:35

Could it be a syntax thing where I code the channels?

**Comment by mdhiggins** [comment]
Looks like your issue is stemming from the fact that you’re using 2 different characters for your single quote and its confusing the JSON object

You have ' and ’

I pasted yours and mine also malfunctioned

If I use the following it works fine, just use the first character for your quote:

data:[8996,4454,514,4454,516,2220,514,2190,544,2218,516,2188,538,2226,514,2190,546,2186,542,2222,510,2218,514,2222,514,2190,546,4454,512,4428,546,4426,546,4424,544],type:'raw'

You don’t need single quotes around the key (data/type) and length does nothing for raw codes

Also you should really try using the encoded version if possible instead of RAW, what kind of remote/TV?

**Comment by kickaxe60** [comment]
I'm with Cogeco cable in Ontario.  With their set top box they use an Atlas
DVR 5-device universal remote.

I pasted your data and still no joy.  Did my base url structure look OK? I
did not get a not supported message though, the dot just flicked around
blue for a few seconds with no feedback.
How are you structuring your verbal command?

I'll keep plugging away.  Thanks again for your help.

**Comment by kickaxe60** [comment]
Also, perhaps my channel up/down syntax is messing things up as this is not
working either... This is the raw syntax I'm using for channel up:
http://xxx.xxx.xxx.x:667/json?pass=pass&plain=[{'type':'raw
','data':[8850,4450,500,4450,450,4400,550,2200,500,4400,500,2200,500,2200,500,2200,550,2150,550,2200,500,2200,500,2200,500,2250,450,4400,550,2200,500,4400,500,2200,500],'khz':38}]

I see now that we must supply a number to increment by for channel up so I
believe my syntax is missing something.

**Comment by mdhiggins** [comment]
Gonna need you to be more specific

The channel increment using JSON you will need to specify a repeat value somewhere, should look like this:

http://xxx.xxx.xxx.x:667/json?pass=pass&plain=[{'type':'raw
','data':[8850,4450,500,4450,450,4400,550,2200,500,4400,500,2200,500,2200,500,2200,550,2150,550,2200,500,2200,500,2200,500,2250,450,4400,550,2200,500,4400,500,2200,500],'khz':38,'repeat':*chan*}]

Also if you could email me your user_id that will be helpful for me to pull the logs and further evaluate

If you say “change TV channel to 50” for example, what response do you get? Does the light come on? If you check your device send history do you see the commands being sent? If the device lights up that’s a good sign it means the command was at least processed

**Comment by kickaxe60** [comment]
I tried adding the ,'repeat':*chan* to the channel up syntax and that did
not work.  When I tell Alexa to “change TV channel to 50” it now says this
device does not support that.  A while ago, it would just blink and say
nothing.  I've sent my amzn1 user id to your email.

**Comment by mdhiggins** [comment]
So interestingly after reviewing the logs it doesn’t look like any of your requests are actually making it to the skill
I see lots of activity on your account but none for adjusting or changing the channel

Looking through your discovery packet it looks like everything should be good to go, but whatever way you’re phrasing things Alexa isn’t even processing the commands to send it to the skill

The official phrasing for channel change is 

“Alexa, change channel to 200 on device”

And the official phrasing for the channel adjustment is

“Alexa, channel up on device”

Try speaking in that format

Also, I notice you’re using the scene controller for input changing instead of the inputController
Were you having issues with that as well? I see HDMI1/HDMI3 

If you set  the input controller and name your inputs hdmi 1 (with a space) that should work natively

**Comment by kickaxe60** [comment]
OK Thanks for spending the time Michael.  I thought things were set up
properly but for some reason, Alexa keeps saying the device does not
support that.  Even the inputs just do not want to work with the TV
device.  I'm pretty sure I tried naming them with and without a space but
I'll try that again to be sure.  I can get channels and inputs to work no
problem using scenes but that is really a cumbersome way to accomplish what
I'm trying to do so at this point, I'm at a loss.  I did try the official
phrasings from the references listed in the controller setup but still no
luck. I'll keep troubleshooting when I get the time.  If you can think of
something else to try, please let me know. I appreciate your help.

**Comment by mdhiggins** [comment]
Keep me posted, I’ll try to think of some other solutions
The other weird thing you can do is try a different name for your device itself, I’ve head Alexa get intermittently confused with just plain ‘TV’ sometimes
The Amazon devs are actively revising and attempting to improve the new API and I’ve seen them break and restore various things during development, they’re still making changes on a daily basis so hopefully things will improve

The other thing to check is in your Alexa history to see exactly what phrasing Alexa is hearing when you say your commands that don’t get sent (IE that device does not support that)

**Comment by kickaxe60** [comment]
Hi Michael, well after a good nights sleep, and a bit of reading on your
GitHub page this morning, I noticed that raw requires kHz data.  Once I put
those in the channel numbers, they worked fine.  I also spaced out HDMI and
1 or 2... and that too worked in the TV input section. The reference
example showed HDMI1 (which is what I used) rather than HDMI 1, so adding
the space fixed it.

Now I'm looking forward to selecting channels by station/network name when
that becomes available.

Thanks again for your help.

Denis

**Comment by mdhiggins** [comment]
I’m actually surprised the kHz thing made a difference, when kHz is not specified it defaults to 38 so unless you needed a different frequency then that shouldn’t have caused issues. But either way I’m glad its work

The reference example for HDMI is wrong on Amazon’s part and I hadn’t realized it either until yesterday. I updated the site to regex check for HDMI# input names and automatically insert the space where appropriate so future users shouldn’t run into that issue

It may be a while for the channel names to work; I’ll need a good free API that contains TV listings to be searchable to link a network to a channel and so far all the ones I can find are paid and require monthly subscription fees from the developer which is out of my price range for this free project. I’ll keep trying to look up and think of alternatives

Glad everything is working for you though!

**Comment by kickaxe60** [comment]
I had a separately named device named COGECO, which is for my cable box,
that I used for the channels and decided to incorporate the channels in my
TV device instead, even though they use different remotes so maybe that had
something to do with it i.e. Alexa likes TV commands and not COGECO when it
comes to channels...

I can easily create scenes for station names for the few that I watch
regularly so that will work in the meantime.  Another cool option would be
to implement guide functions within the device but that too I can use
scenes for to navigate...

Thanks again.

**Comment by mdhiggins** [comment]
Well the ChannelController API supports network names, so saying something like “change channel to ABC” should work
The caveat is that the backend code needs some way to translate ABC into a numerical channel on a per user basis

One option I’m considering is rather than depending on a specific API is to allow users to create their own channel map and have it just contact the pythonanywhere server for that data and keep on the user to create a map of networks they care about

I’ll have to play around with it and try and find the best solution

**Comment by kickaxe60** [comment]
That's kind of what I'm doing with scenes.  I'm calling each one by channel
name e.g. TSN1 and firing off the numerical codes in sequence to tune into
that channel.  So far it seems to work fine. It's a bit of a pain but you
only have to do each one once as the channel numbers rarely change.

**Comment by mdhiggins** [comment]
Just pushed a big update to the setup site
You can now add more than 4 input options via the input controller
But more importantly.. I added support for channelController stations
You can define your stations and their channel numbers now and then verbally change channels based on those names; please give it a try and let me know if its working

Mike

**Comment by kickaxe60** [comment]
Will check it out tomorrow. 
Thanks

**Comment by kickaxe60** [comment]
When I click on Define your channels I get:
Error!
There was an error processing your saved JSON object, it is likely
corrupted or invalid and will need to be recreated. Scroll to the bottom of
this page and try to salvage it.

I'm a little hesitant to try this until I'm sure I won't lose what I've
done so far...

**Comment by mdhiggins** [comment]
Whoops, that error is an accident, should be the welcome message
Fixed now

Additionally, you can backup your whole config but copying the JSON object to a text file on the main page at the bottom, if ever there’s a problem pasting that back in that location will restore all your devices and settings

Let me know how it goes
Mike

**Comment by kickaxe60** [comment]
OK I have made a backup of my JSON file.  I will play with this tomorrow
morning and let you know how I make out. Were there changes to the ino
sketch or just the controller setup?

**Comment by mdhiggins** [comment]
Just the setup and Alexa backend. No ino changes

**Comment by kickaxe60** [comment]
Also, any particular way the voice command needs to be structured for Alexa?

**Comment by mdhiggins** [comment]
Nah it’s structured pretty much as you’d expect. I’m gonna have to add some extra logic to try and reduce errors depending on how Alexa formats the channel name but not uncomplicated common stations it’s working well.

**Comment by kickaxe60** [comment]
When I add a station name and number and go to update channels, I get the
same corruption error...

**Comment by mdhiggins** [comment]
Try now. Should be fixed.

**Comment by kickaxe60** [comment]
Hi Michael, it's very hit and miss at this point.  I set up 14 of them and
only three work as expected. The first one is called cogeco on 700 and it
recognizes cogeco but only does a channel up.  The next two, ctv and cbc
are not recognized (Alexa responds: that command doesn't work on device
tv). I created a scene called cbc for channel 702 and it works flawlessly.
The next two, city and global work as expected.  The next  8 channels,
chch, omni1, omni2, abc, cbs, nbc, pbs, and fox, get the same error from
Alexa. The last one called tsn 1 works as expected.  I tried with and
without spaces on the lettered networks and that made no difference.  The
command that I use is: Alexa, change TV channel to global. I've tried a few
variations for those that don't work with no luck...

**Comment by mdhiggins** [comment]
Alright I updated my algorithm for matching channels to be more robust with ways Alexa may interpret voice names for channels
I actually copied your channel data and tested them all now and they all appear to be recognized

Cogeco is the only one that is iffy, you have to pronounce it with a ‘j’ sound cojecko without an accent on the final ‘co’ and then it recognizes it

Give it a shot

**Comment by kickaxe60** [comment]
Still have issues with ctv, chch, and cogeco.  Omni1 & 2 work sometimes.
I'll keep plugging away.  Is there a limit I should be worried about?

**Comment by mdhiggins** [comment]
No limit to worry about

I notice you changed your channel names and added spaces, the best formatting is one with no spacing between letters (ABC, NBC etc). The algorithm strips out spaces and punctuation created by Alexa to try to match with something in your list

Using the following all of these work except cogeco just cause Alexa interprets things weird for that one, might need to contact Alexa since that’s apparently a big Canadian network they should recognize

[{
		"name": "cogeco",
		"channel": "700"
	}, {
		"name": "ctv",
		"channel": "701"
	}, {
		"name": "cbc",
		"channel": "702"
	}, {
		"name": "city",
		"channel": "704"
	}, {
		"name": "global",
		"channel": "706"
	},
	{
		"name": "chch",
		"channel": "707"
	}, {
		"name": "omni 1",
		"channel": "708"
	}, {
		"name": "omni 2",
		"channel": "709"
	}, {
		"name": "abc",
		"channel": "712"
	}, {
		"name": "cbs",
		"channel": "713"
	}, {
		"name": "nbc",
		"channel": "714"
	}, {
		"name": "pbs",
		"channel": "715"
	}, {
		"name": "fox",
		"channel": "716"
	}
]

**Comment by kickaxe60** [comment]
OK, I'll remove the spaces. I'll maybe change the cogeco channel to
something else. I've added a bunch more but have not tested them.  Won't
likely get to that until tomorrow and will let you know how I make out.

Great work on this by the way.  My wife loves not having to fiddle with
remotes.

**Comment by kickaxe60** [comment]
Hi Michael, I pretty much have all of the station names working now.  It
would be nice to have some guide functionality built in too e.g. turn on
guide, show favourites, page up/down, scroll up/down, show info, select,
last channel, exit...

I also have a bit of an issue with one of the PVR commands.  I can say
Alexa, Rewind/Fast Forward/Pause/Stop PVR but Play PVR results in Alexa
looking to play music by a band named PVR in amazon music.  If I say Alexa,
PLAY ON the PVR it works but it's a bit awkward.

Thanks for your help.  Keep me posted on any other enhancements, I'd be
glad to test them out for you.

Take care,

Denis

**Comment by mdhiggins** [comment]
There’s no guide functionality built into the API so for now that’s not possible. Will require amazon support to expand that. 

The play issue is known and is another Alexa issue unrelated to my skill. They seem to be actively tweaking their algorithm to improve this so hopefully it will get better. 

Glad the channel controller is working better now though!

**Comment by kickaxe60** [comment]
Not a biggie re: the guide as I have the eight or so commands set up as
scenes and they work fine.

**Comment by kickaxe60** [comment]
Hi Michael, just wondering if you're updating something.  Alexa keeps
saying that TV is not responding.  Thought I would check with you before I
try troubleshooting...

**Comment by kickaxe60** [comment]
Never mind, it was a loose wire on my breadboard. I will need to solidify
that :)

**Comment by mdhiggins** [comment]
Hah no worries. I just started soldering all mine using Raptors PCB it’s pretty great

**Comment by kickaxe60** [comment]
Hi Michael, just curious if something was updated on your end or if perhaps
Amazon has changed something.  As of this morning, every voice command I
issue results in the dot whirring in the blue circle for about 5 seconds
and then Alexa says "sorry, *device name* is not responding" but then the
command is triggered any way after her response.  It works but there is lag
and a negative response from Alexa.  I haven't changed anything and it
worked perfectly fine yesterday so I thought I would let you know.

**Comment by mdhiggins** [comment]
Didn't change anything on my end
Looking at the amazon logs, it looks like those failures was your
connection requests timing out
I had this happen once and it was due to a router issue and restarting the
router and rebooting the devices fixed it (the IP addresses were double
booked and the devices would receive the commands but be unable to respond
to Alexa to confirm the command was successful received).

**Comment by kickaxe60** [comment]
Rebooted both the router and the echo dot and still does the same thing.
Odd since I haven't changed anything...  Anything else you can think of
that I can try?

**Comment by kickaxe60** [comment]
Also, I can control my smart plugs no problem via Alexa.  It's just the IR
Controller stuff that started acting up this morning...

**Comment by mdhiggins** [comment]
Try changing the IP address of your device and updating the port forwarding to the new IP and see if it works. 

Can you ping your device or access the status page from an external address?

**Comment by kickaxe60** [comment]
Hi Michael, out of the blue it started functioning normally about 2 hours
ago.  I hadn't tried anything since this morning when I rebooted everything
but still wasn't working... Thanks for the IP/port forwarding tip.  If it
happens again, I'll add that to my troubleshooting steps.

Take care,

Denis

**Comment by mdhiggins** [comment]
Glad it’s working

**Comment by kickaxe60** [comment]
Hi Michael, this has been working great now for quite a while.  I've
decided to add another one to our other TV.  It seems that my setup doesn't
like two ESPs working at the same time for some reason.  They are set up
with different names, IP addresses, port forwarding, and passwords.  When
one works, the other doesn't, i.e. I'll unplug/replug the one that doesn't
work and that gets it going again but then it messes the other one up.  Is
there something I'm missing?  Is there anything else that needs to be done
on the software side of the arduino?

Thanks

Denis

**Comment by kickaxe60** [comment]
Hi Michael, well I just reset them both and now they are working.  No
explanation as I didn't do anything different today.  It appears then that
they are forgotten by the network on occasion until I restart them.  Would
that be an arduino issue or something to do with my router?

Thanks

Denis

**Comment by mdhiggins** [comment]
Hard to say but if its working it might have just been a router or
networking issue

**Comment by kickaxe60** [comment]
Hi Michael, ever since I added the second IR controller, it interferes with
the first one somehow.  If I reset them, they seem to work for a little
while then one or both no longer respond later on... I've got many devices
running on my wifi without issue so I can't pinpoint the problem, but
definitely had no issues before adding the second one.   Right now I can
ping the first one no problem but the second one currently doesn't respond,
so it has gone offline somehow.  Let me know if you can think of anything
else to try.

Thanks

Denis

**Comment by kickaxe60** [comment]
I should also mention that the one that currently does not respond to the
ping, gets discovered when I run discover devices...

**Comment by kickaxe60** [comment]
Hi Michael, I did some googling and came across this thread:
https://github.com/esp8266/Arduino/issues/2330

Many of the posts are about users exhibiting the issues I have.  There is
debate as to whether it's router related or ESP.  Some issues relate to
single ESPs while others to multiple ESPs. Some problems were fixed by
firmware upgrade to TP-Link routers however, I have a Netgear Nighthawk
(R7000) with the most up-to-date stock firmware and I have had zero issues
with my other devices.  Some of the posters have been successful in
resolving their issue by adding "WiFi.setSleepMode(WIFI_NONE_SLEEP)" before
"WiFi.begin()" in the code however, I don't see the "WiFi.begin()" command
anywhere in your code.  Is it possible to achieve this somewhere in your
code i.e. prevent the NodeMCU ESP8266 devices from ever going to sleep?
Both of my ESPs work from power so I am not concerned with battery life...

Thanks

Denis
